### PR TITLE
Add ability to avoid accumulating multiple search results

### LINKED ISSUE #8777: [Feature Request] Option to remove old search results from Find-result window when doing a new search
There as been a user request to clear out any prior-search results from the _Find result_ window when a new search that writes output to that window is initiated.

See https://community.notepad-plus-plus.org/topic/19850/setting-to-clear-previous-search-results-before-files-search

This seems to be a reasonable idea, as a subsequent right-click menu _Open All_ can easily (and unintentionally) open many more documents than the user truly needs/wants (this was pointed out by a different user in the above Community thread).

### Debug Information
Feature request based on behavior current to N++ version 7.8.9.

### PR DESCRIPTION
**Pull request by sasumner**
Fix #8777

So this adds the mentioned capability without adding any new GUI elements.
It adds a "latching effect" to the _Clear all_ right-click menu item in the _Search results_ window.

If you use the command when the window is already empty, it will put a checkmark on the _Clear all_, indicating that the window is always to be cleared before new output is written.  In this way, old searches won't accumulate.

The behavior persists between runs of Notepad++, so if this is what you want permanently, you'd set the checkmark one time and forget about it for the future.

I like the UI design, which of course would have to be documented, but I realize that it might not be to the liking of everyone.

### CODE REVIEW

**Comment by ArkadiuszMichalski** [comment]
@sasumner 
I can't open Search Results Window without making one search. It's intentional? Even when I delete all the results, I can open them and it is empty. So why it's blocked at first time?

**Comment by sasumner** [comment]
>I can't open Search Results Window without making one search. It's intentional?

Guess so.  That's the way it has always been.

**Comment by mere-human** [comment]
A bit off-topic, but I see this from time to time when doing a search in files using a debug version.
Has anyone seen this? I thought that it might be related to binary files in the list. But I'm not sure.
image

**Comment by sasumner** [comment]
> Has anyone seen this?

Yes, I've seen that.  I think it has to do with the _Search results_ lexing happening on partially populated data. but to be honest I'm not 100% sure of that.

File: PowerEditor/src/ScintillaComponent/FindReplaceDlg.cpp [hunk 1]
```diff
@@ -4086,7 +4093,29 @@ INT_PTR CALLBACK Finder::run_dlgProc(UINT message, WPARAM wParam, LPARAM lParam)
...
+							_clearBeforeEverySearch = true;
+						}
+						removeAll();
+					}
+
+					if (!_canBeVolatiled && _clearBeforeEverySearch != old)
+					{
+						// only remember this setting from the original finder
+						NppParameters& nppParam = NppParameters::getInstance();
+						NppGUI& nppGUI = const_cast<NppGUI&>(nppParam.getNppGUI());
```
[the remark this answers is not in the document]
Comment (sasumner) [reply PowerEditor/src/ScintillaComponent/FindReplaceDlg.cpp:4115]: It's rather boilerplate code, copied from the way it is done in many other places.  Thus, TBH I haven't thought too much about it.

**Comment by sasumner** [comment]
>So, in order to clear all always a user have to do these steps:

>Perform some search in order to open "Search results" window.
>Select "Clear all" from context menu to clear any results
>Select "Clear all" again to clear before a new search operation.
>Not very intuitive. But better than nothing.

Aside from a preference setting, yes, you'd always have to perform a search first, before being able to get to the context menu.
This is existing behavior right now, say if you wanted to set the setting to wrap long lines.

>I've read a long discussion in the linked issue.
>It seems that a new preference is discouraged. Ok, understandable.

Not exactly discouraged, but probably the goal is to avoid the _Preferences_ looking like the "airplane cockpit"

>How about a new context menu entry "Clear always"?

Sure, can be done, but maybe in keeping with existing verbiage maybe it should be "Purge for each search" ? :-)
Minor right now:  When does the context menu start looking like that "airplane cockpit"?

My idea was to add a new feature that is likely to be used one time by a user, in an "out of the way" place.
There will probably be two camps, those that like existing behavior (hate change?) and those that will always want the new behavior.

>To me, it would be more obvious. Because now we are solving 2 different problems with one menu item.

We haven't solved anything yet, that's why I want feedback. :-)

**Comment by sasumner** [comment]
Another idea might be to change the menu item to be "Clear (+Shift to clear before each search)" but that has wordiness problems.

Perhaps I'm stuck on the idea to not "waste" a second menu item when the functionality is related enough that the two things would never be used "together" -- meaning that if you have the "keep clear" functionality on, you'd never do a simple "clear".  OK, you _could_ do it, but to what purpose?

**Comment by ArkadiuszMichalski** [comment]
@sasumner 
>...meaning that if you have the "keep clear" functionality on, you'd never do a simple "clear". OK, you could do it, but to what purpose?

With your solution its still possible, just use Delete key, but why delete only 1 entry and have an empty Search Results window? Rather I would except that the Search Results window would be opened without doing the first search. This lock made sense as there were no toggle options in the context menu, but now there's one (and maybe yours will  be add, who knows if no others will come in the future, like ["new window for each search"](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8777#issuecomment-683644297)). Why do I have to be forced to do dummy searches to change the behavior of this window?

**Comment by sasumner** [comment]
@ArkadiuszMichalski 

> I would expect that the Search Results window could be opened without doing the first search.

>Why do I have to be forced to do dummy searches to change the behavior of this window?

I would rather close this PR and forget about your desired functionality regarding persistent-clear, if we're going to force it to consider why the _Search results_ window can't be shown before the first search is run.  :-)

For myself, I think it is a very small requirement to have to run a search before adjusting options.
Because:  No one can even remember how things are currently set BEFORE seeing a search result with those effects in it (e.g. "Oh, yea, these lines that I see displayed here aren't wrapped, I need to turn that on...").

But, it seemed you didn't mind my "creative"(?) design for "locking" the persistent clear?
To me I like it (or I wouldn't have put it forth) but I do understand that not all would like it, and I'll certainly change it when the "best" way of doing it emerges from these discussions.

**Comment by ArkadiuszMichalski** [comment]
@sasumner 
> But, it seemed you didn't mind my "creative"(?) design for "locking" the persistent clear?

Because I'd rather have it this way than none. If this option is enabled, I would only change the description from "Clear all" to "Clear all (except first)" or something like that, because it doesn't actually delete all of them.

**Comment by sasumner** [comment]
Any of these make a shared functionality of one menu item more palatable?

image

image

image

**Comment by mere-human** [comment]
> Any of these make a shared functionality of one menu item more palatable?

As for me, it only adds more confusion. Especially the fact that the "clear" word is duplicated.

> Minor right now: When does the context menu start looking like that "airplane cockpit"?

The context menu in "Search results" window is still smaller than the one in the document area.
If it becomes too big, second level sub-menus could be added.

How about "✔Keep results"?🙂

**Comment by sasumner** [comment]
>>...start looking like that "airplane cockpit"?

The problem with it is it tends to creep up on you...one addition is fine, then another is ok, then WHAMMO, you're at the cockpit and you don't know how you got there.  :-)

>The context menu in "Search results" window is still smaller than the one in the document area.
If it becomes too big, second level sub-menus could be added.

Agreed...OK....

Since this feature has to have its own separate menu item, I think there isn't much room for choice; it has to be:

`Purge for each search`

:)

Given the continuing feedback, I will make this change to this PR.

**Comment by sasumner** [comment]
Latest incarnation of the UI:

image

Maybe it's me, but it is feeling like a large context menu now.  :-)

**Review by donho — CHANGES_REQUESTED** [review]
Solve the conflict.

**Review by donho — APPROVED** [review]

**Comment by sasumner** [comment]
Documentation to be updated per https://github.com/notepad-plus-plus/npp-usermanual/issues/184

**Pull request closed without merging**